### PR TITLE
Add upload file size check to Joomla Update Component Upload & Update

### PR DESCRIPTION
**Pull request by richard67**
Pull Request for Issue #27494 .

### Summary of Changes

This Pull Request (PR) is a back port of PR #27570 from 4.0-dev.

It adds following to the Upload & Update tab of the Joomla Update Component:
- Display of the selected upload file's size.
- Check if the selected file is too big to be uploaded with respect to the PHP limits.

If the selected upload file is too big, a warning alert is shown. If the user tries to upload that file anway, the upload is rejected with an appropriate browser alert message.

The size of the selected file is always shown in MB because the PHP limits and so the maximum file size will with a 99.9 % likelyhood also be in MB.

For the Upload & Install Package tab of the Extension Installer I've created PR #30190 for backporting PR #27665 from 4.0-dev.

### Testing Instructions

1. Go to `administrator/index.php?option=com_joomlaupdate` and there to tab "Upload & Update".

2. Try to upload a file which is bigger than the maximum size shown below the file select button.
**Hint:** This does not necessarily need to be a Joomla Update paclage. Take any file with a sufficient size.
**Result:** See section "Actual result BEFORE applying this Pull Request" below.

3. Apply patch of this PR.

4. Go to `administrator/index.php?option=com_joomlaupdate` and there to tab "Upload & Update".

5. Check the `Joomla package file` fields.
**Result:** See section "Expected result" screenshot 1. No information related to the selected file's size is shown if no file is selected.

6. Select a file which is **not** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 2. The size of the selected file is shown below the maximum upload size.

7. Select a file which **is** too big compared to the maximum upload size shown below the file selection.
**Hint:** This can be any file, it doesn't necessarily have to be a Joomla Update package.
**Result:** See section "Expected result" screenshot 3. The size of the selected file is shown below the maximum upload size, and below that a warning alert is shown telling that the selected file is too big.

8. Try to upload the file anway by selecting the `Upload & Install` button.
**Result:** See section "Expected result" screenshot 4. A browser alert is shown telling that the file can't be uploaded because it's too big. The upload is not started.

9. Toggle the file selection a few times between too big and not too big file, and also between different files of each type.
**Result:** The warning disappears when the selection is changed from too big to not too big, and it appears again when vice versa. The file size is updated with each change of the file.

10. Finally select a valid Joomla Update package which is not too big and verify that uploading still works.
**Result:** Uploading of not too big file still works as well as without this PR.

### Actual result BEFORE applying this Pull Request

The result depends on the values of the PHP limit `post_max_size` and `upload_max_filesize`.

- If `post_max_size` is big enough but `upload_max_filesize` is violated, you get an error alert with the untranslated language string constants `COM_INSTALLER_MSG_INSTALL_WARNINSTALLUPLOADERROR` and `COM_INSTALLER_MSG_WARNINGS_SMALLUPLOADSIZE`.
The not translated languages strings will be solved with PR #29899 .

- If `upload_max_filesize` is big enough but `post_max_size` is violated, you get stuck in the update process, i.e. nothing happens, no progress.

### Expected result AFTER applying this Pull Request

- No file selected:
expected-result_no-file-selected

- File size ok:
expected-result_file-size-ok

- File too big:
expected-result_file-size-too-big

- Broswer alert when trying to upload too big file:
expected-result_browser-alert

### Documentation Changes Required

None.

### CODE REVIEW

File: administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php [hunk 1]
```diff
@@ -96,9 +125,15 @@
...
 				</td>
 				<td>
-					<input class="input_box" id="install_package" name="install_package" type="file" size="57" accept=".zip,application/zip" /><br>
-					<?php $maxSize = JHtml::_('number.bytes', JUtility::getMaxUploadSize()); ?>
-					<?php echo JText::sprintf('JGLOBAL_MAXIMUM_UPLOAD_SIZE_LIMIT', '&#x200E;' . $maxSize); ?>
+					<input class="input_box" id="install_package" name="install_package" type="file" size="57" accept=".zip,application/zip" onchange="Joomla.installpackageChange()" /><br>
+					<?php $maxSizeBytes = JUtility::getMaxUploadSize(); ?>
+					<?php $maxSize = JHtml::_('number.bytes', $maxSizeBytes); ?>
+					<input id="max_upload_size" name="max_upload_size" type="hidden" value="<?php echo $maxSizeBytes; ?>" />
+					<div class="small"><?php echo JText::sprintf('JGLOBAL_MAXIMUM_UPLOAD_SIZE_LIMIT', '&#x200E;' . $maxSize); ?></div>
```
Comment (Quy) [inline administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:132]: `&#x200E;` This character does not display. What does it look like?

Comment (richard67) [reply administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:132]: @infograf768 Can you answer @Quy 's question? I can't, I've just copied it as it was for the maximum size display.

Comment (richard67) [reply administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:132]: Maybe it's a kind of non-breaking space?

Comment (wojsmol) [reply administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:132]: @Quy This is left-to-right mark - see [this](https://www.w3.org/TR/WCAG20-TECHS/H34.html) W3C article for details.

Comment (richard67) [reply administrator/components/com_joomlaupdate/views/default/tmpl/default_upload.php:132]: @Quy If ok with the above explanation, mark this conversation as resolved. Thanks in advance.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f9a655c98d374a653f2668ec0a168e822723c7d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29895">issues.joomla.org/tracker/joomla-cms/29895</a>.</sub>

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on f9a655c98d374a653f2668ec0a168e822723c7d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29895">issues.joomla.org/tracker/joomla-cms/29895</a>.</sub>

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on f9a655c98d374a653f2668ec0a168e822723c7d3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29895">issues.joomla.org/tracker/joomla-cms/29895</a>.</sub>

**Comment by jwaisner** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/29895">issues.joomla.org/tracker/joomla-cms/29895</a>.</sub>

**Comment by richard67** [comment]
@Quy @jwaisner Thanks a lot for testing. There is another, similar PR for the extensions installer: #30190 . It doesn't show the size of the selected file like this one here does, because in the extension installer's Upload & Install the upload starts immediately after having selected or dropped a file, but the alert if the file is too big is the same as here. So if you could test the other one, too, would be really great. Thanks in advance.

**Comment by HLeithner** [comment]
Thanks

**Comment by richard67** [comment]
Thanks.

**Pull request merged**